Make ocsp optional updates

ryancdickson opened the pull request:
Staging updates for continued discussion....

This PR is intended to make it easier to collect community feedback.

ryancdickson committed:
Update BR.md

ryancdickson committed:
Update BR.md

ryancdickson committed:
Update BR.md

ryancdickson committed:
Improve specificity of CRL issuance frequency

ryancdickson committed:
Update BR.md

ryancdickson committed:
Update BR.md

ryancdickson committed:
Update BR.md

ryancdickson committed:
Update BR.md

ryancdickson committed:
Update BR.md

ryancdickson committed:
Update BR.md

ryancdickson committed:
Update BR.md

aarongable commented on docs/BR.md:
```diff
@@ -1286,29 +1286,36 @@ No stipulation.
+
+Within 24 hours of issuing its first Certificate, CAs MUST generate and publish either:
+- a full and complete CRL; OR
 - partitioned (i.e., "sharded") CRLs, that when aggregated, represent the equivalent of the full and complete CRL.
```
nit:
```suggestion
- partitioned (i.e., "sharded") CRLs that, when aggregated, represent the equivalent of a full and complete CRL.
```

aarongable commented on docs/BR.md:
```diff
@@ -1286,29 +1286,36 @@ No stipulation.
-- the value of the `nextUpdate` field MUST NOT be more than twelve months beyond the value of the `thisUpdate` field.
+CAs MUST continue issuing CRLs until one of the following is true:
+- all CA Certificates containing the same Subject Public Key are expired or revoked, OR
+- the corresponding CA Private Key is destroyed
```
nit:
```suggestion
- the corresponding CA Private Key is destroyed.
```

ryancdickson committed:
Update BR.md

Typo (thanks, Wendy!)

aarongable commented on docs/BR.md:
```diff
@@ -1286,29 +1286,36 @@ No stipulation.
+
+| __Conditions__ | __Minimum CRL Issuance Frequency__ | __Maximum nextUpdate Period__ | 
+| ----           | -                                  | -                             |
+| The CA has revoked at least one unexpired Certificate since the `thisUpdate` field of the last CRL. | A CRL MUST be generated and published within 24 hours of the `thisUpdate` field of the last CRL. <br><br> A CRL SHOULD be generated and published within 4 hours of the `thisUpdate` field of the current CRL. | The value of the `nextUpdate` field MUST NOT be more than 10 days beyond the value of  ...
```
I must admit I'm still not enamored with this use of "has revoked". Despite Dimitris' totally valid point that we already do this for Subordinate CA Certificates, to me it feels like an unfortunate precedent to carry forward. How can a CA be said to "have revoked" something if they haven't published a CRL containing it yet?

Also, this phrasing is impossible. Suppose that a CA issues a CRL, and then coasts for 5 days with no revocations being necessary. This is permitted under the "All other conditions" entry. Then, a cert needs to be revoked. But significantly more than 24 hours have already passed since the `thisUpdate` of the previous CRL! So in practice, a CA would be required to issue at least every 24 hours, just in case a new revocation request comes in.

These kinds of accidental gotchas are why I'm opposed to a two-tiered system of CRL issuance frequency requirements. We've already demonstrated that it is difficult to write and read (I missed this contradiction on my first two reads of this diff!) requirements like this. We should not be making things more difficult than necessary.

I'm still in favor of the previous "CAs issuing Subscriber Certificates SHALL update and reissue CRLs at least once every 24 hours".

ryancdickson replied:
Abandoned the table (the more I thought about it, the more it seemed to overcomplicate things).

[Back](https://github.com/ryancdickson/staging/commit/260d81518b8df59aa55577f4b52650186ced2ff5) to a bulleted list and including a slightly modified version of Dimitris' latest proposal.

dzacharo replied:
> I must admit I'm still not enamored with this use of "has revoked". Despite Dimitris' totally valid point that we already do this for Subordinate CA Certificates, to me it feels like an unfortunate precedent to carry forward. How can a CA be said to "have revoked" something if they haven't published a CRL containing it yet?
> 
> Also, this phrasing is impossible. Suppose that a CA issues a CRL, and then coasts for 5 days with no revocations being necessary. This is permitted under the "All other conditions" entry. Then, a cert needs to be revoked. But significantly more than 24 hours have already passed since the `thisUpdate` of the previous CRL! So in practice, a CA would be required to issue at least every 24 hours, just in case a new revocation request comes in.
> 
> These kinds of accidental gotchas are why I'm opposed to a two-tiered system of CRL issuance frequency requirements. We've already demonstrated that it is difficult to write and read (I missed this contradiction on my first two reads of this diff!) requirements like this. We should not be making things more difficult than necessary.
> 
> I'm still in favor of the previous "CAs issuing Subscriber Certificates SHALL update and reissue CRLs at least once every 24 hours".

I believe this reading is a bit egregious. Every CA must record the status of a certificate (intermediate or end-entity). Due to the way PKI works, there is always some delay (greater than zero), between changing the status of a certificate to "revoked" and issuing a CRL or signing a new OCSP response. This is especially obvious when CAs perform offline ceremonies with CRLs/OCSP responses issued from Root CAs. I hope people agree with this.

I [proposed some text](https://github.com/ryancdickson/staging/pull/3#discussion_r1186875190) that I consider clear and unambiguous. I hope more Members can chime-in and state their opinion whether this language works and delivers what is expected.

ryancdickson replied:
(Reviewed and incorporated @dzacharo suggested edits.)

aarongable commented on docs/BR.md:
```diff
@@ -1286,29 +1286,36 @@ No stipulation.
+| __Conditions__ | __Minimum CRL Issuance Frequency__ | __Maximum nextUpdate Period__ | 
+| ----           | -                                  | -                             |
+| The CA has revoked at least one unexpired Certificate since the `thisUpdate` field of the last CRL. | A CRL MUST be generated and published within 24 hours of the `thisUpdate` field of the last CRL. <br><br> A CRL SHOULD be generated and published within 4 hours of the `thisUpdate` field of the current CRL. | The value of the `nextUpdate` field MUST NOT be more than 10 days beyond the value of  ...
+| All other conditions. | A CRL MUST be generated and published within 7 days of the `thisUpdate` field of the last CRL. <br><br> A CRL SHOULD be generated and published within 4 hours of the `thisUpdate` field of the current CRL. | The value of the `nextUpdate` field MUST NOT be more than 10 days beyond the value of the `thisUpdate` field. |
```
Is it intentional that the SHOULD here is also 4 hours, and not 24 hours as suggested in Dimitris' latest message?

ryancdickson replied:
See above (abandoned this approach).

Justification for 4 hours was because I misunderstood discussion on thread. I thought people were interpreting Apple's policy ("_Under normal operating conditions, the CRL URLs provided by CAs in this section must be available such that Apple systems are able to successfully retrieve the current CRL every 4 hours._") to mean CRLs have to be updated every 4 hours (not the case).

aarongable commented on docs/BR.md:
```diff
@@ -1286,29 +1286,36 @@ No stipulation.
+CAs issuing CA Certificates MUST issue CRLs in accordance with the table below:
+| __Conditions__ | __Minimum CRL Issuance Frequency__ | __Maximum nextUpdate Period__ | 
+| ----           | -                                  | -                             |
+| The CA has revoked at least one unexpired Subordinate CA Certificate since the `thisUpdate` field of the last CRL. | A CRL MUST be generated and published within 24 hours of the `thisUpdate` field of the last CRL. <br><br> A CRL SHOULD be generated and published within 4 hours of the `thisUpdate` field of the current CRL. | The value of the `nextUpdate` field MUST NOT be more than twelve months ...
```
As above, this phrasing is also impossible. If a Root CA issues a 12-month CRL, then 6 months pass, then it needs to revoke one of its Subordinate CA Certificates, it is stuck: it "has revoked" (see comment above) at least one cert since the thisUpdate of the previous CRL, but significantly more than 24 hours have already passed since that previous CRL.

dzacharo replied:
Please see https://github.com/ryancdickson/staging/pull/3#discussion_r1186880435.

ryancdickson replied:
I'll mark this comment as resolved given the "table" presentation format was later abandoned. I included @dzacharo suggested edits on 4.9.7 and think that it reads well.

Any further comments or concerns on the current 4.9.7 text can be discussed in the open comment above.

aarongable commented on docs/BR.md:
```diff
@@ -1321,9 +1328,7 @@ defined by RFC6960.
-The CA MAY support on-line revocation status checking using OCSP.
-
-If the CA supports OCSP, the following requirements SHALL apply:
+If the CA signs OCSP responses either directly or through a delegated OCSP responder, the following requirements SHALL apply.
```
I do like this phrasing update. However, it doesn't fully answer my question:

Is it acceptable for a CA to operate an OCSP service for only *some* of the non-expired certificates it has issued?

For example, a CA might switch from including an AIA OCSP url in its certificates, to including a CRLDP url in its certificates, with no overlap: each Subscriber cert contains only one or the other. Is the CA required to provide OCSP responses, and update those responses no later than four days after the thisUpdate, even for the certificates that do not embed an AIA OCSP url?

The previous phrasing ("support on-line revocation status checking using OCSP") and the current phrasing ("signs OCSP responses") both seem to assume that OCSP support is a binary thing: either the CA does it, or it doesn't. This makes sense both in today's world when OCSP is required, and in the potential future when OCSP is wholly optional. It raises questions about the transition period, however.

To be clear, I don't think this is insurmountable. For example, a CA could choose to stop including AIA OCSP urls, but continue providing OCSP responses for all certificates until all certificates with an AIA OCSP url have expired, then instantaneously stop providing any OCSP responses at all. But since many CAs will be going through a transition period here, some clarity if possible would be nice.

CBonnell replied:
In the CP for another PKI that is based on the BRs and EVGs, we made OCSP optional for Subscriber Certificates. To make it abundantly clear that the OCSP requirements were relevant only to those certificates that the CA "intended" to support OCSP, we used this turn of phrase:

```
For the status of Subscriber Certificates which include an Authority Information Access extension with a id-
ad-ocsp accessMethod (“AIA OCSP pointer”)
```

We may want to do similar here.

ryancdickson replied:
@CBonnell, thanks for sharing!

@aarongable / @CBonnell: Any concerns / recommended improvements for ... "_The following SHALL apply for communicating the status of Certificates which include an Authority Information Access extension with an id-ad-ocsp accessMethod (“AIA OCSP pointer”)_."

CBonnell replied:
LGTM

aarongable replied:
I like this approach

ryancdickson committed:
Update docs/BR.md

Editorial

Co-authored-by: Aaron Gable <aaron@aarongable.com>

ryancdickson committed:
Update docs/BR.md

Editorial

Co-authored-by: Aaron Gable <aaron@aarongable.com>

ryancdickson committed:
Update BR.md

ryancdickson committed:
Update BR.md

ryancdickson committed:
Update BR.md

Address comment from Aaron: "I'm not in favor of allowing CRLs to remain non-updated for 7 days because that is a regression from current OCSP behavior. Section 4.9.10.(4) makes it so that updated revocation information is always available "no later than four days after the thisUpdate". Therefore, a CA operating in a CRLs-only mode should be required to update their CRLs at least once every 4 days."

ryancdickson committed:
Update BR.md

ryancdickson committed:
Update BR.md

ryancdickson committed:
Update BR.md

ryancdickson committed:
Update BR.md

ryancdickson committed:
Update BR.md

ryancdickson committed:
Update BR.md

ryancdickson committed:
Update BR.md

ryancdickson committed:
Update BR.md

ryancdickson committed:
Update BR.md

ryancdickson committed:
Update BR.md

dzacharo commented on docs/BR.md:
```diff
@@ -1286,29 +1286,35 @@ No stipulation.
+CAs MUST continue issuing CRLs until one of the following is true:
+- all CA Certificates containing the same Subject Public Key are expired or revoked; OR
+- the corresponding CA Private Key is destroyed.
 
```
Tried to use similar language for Subscriber and CA Certificates for consistency. I also proposed following the NSR style for numbers, especially now that we have so many different numbers in the same subsection. It would be great if we could make a pass through the entire document and apply this style in a future cleanup ballot.

```suggestion
CAs issuing Subscriber Certificates:  
1. MUST update and publish a new CRL within at least: 
     - seven (7) days if all Certificates include an Authority Information Access extension with an id-ad-ocsp accessMethod (“AIA OCSP pointer”); or
     - four (4) days in all other cases; 
2. MUST update and publish a new CRL within twenty four (24) hours after recording a Certificate as revoked; and
3. MUST include a `nextUpdate` field value that is no more than ten (10) days beyond the value of the `thisUpdate` field.

CAs issuing CA Certificates:  
1. MUST update and publish a new CRL within at least twelve (12) months;
2. MUST update and publish a new CRL within twenty four (24) hours after recording a Subordinate CA Certificate as revoked; and
3. MUST include a `nextUpdate` field value that is no more than twelve (12) months beyond the value of the `thisUpdate` field.

CAs MUST continue issuing CRLs until one of the following is true:
- all Subordinate CA Certificates containing the same Subject Public Key are expired or revoked; OR
- the corresponding Subordinate CA Private Key is destroyed.
```

ryancdickson replied:
Thank you @dzacharo. 

I think standardizing format and style would be really helpful, and possibly something we could target for the cleanup ballot.

Some of the items we could standardize:
- bulleted list conventions
- table conventions
- use of numbers
- use of examples

dzacharo replied:
Filed https://github.com/cabforum/servercert/issues/432.

dzacharo commented on docs/BR.md:
```diff
@@ -2905,11 +2909,11 @@ This section contains several fields that are common among multiple certificate
 The CRL Distribution Points extension MUST be present in:
 - Subordinate CA Certificates; and
-- Subscriber Certificates with validity greater than ten days.
+- Subscriber Certificates with 1) validity greater than ten days and that 2) do not include an Authority Information Access extension with an id-ad-ocsp accessMethod.
```
We should probably avoid using the exact number of days because the definition of "Short-lived Subscriber Certificate" is supposed to take care of that. As much as I hate "negative" references (non-short-lived certificates), I don't see an easy way to avoid it.

```suggestion
- Subscriber Certificates that 1) do not qualify as "Short-lived Subscriber Certificates" and 2) do not include an Authority Information Access extension with an id-ad-ocsp accessMethod.
```

ryancdickson committed:
Update docs/BR.md

Co-authored-by: Dimitris Zacharopoulos <dzacharo@users.noreply.github.com>

ryancdickson committed:
Update docs/BR.md

Co-authored-by: Dimitris Zacharopoulos <dzacharo@users.noreply.github.com>

ryancdickson committed:
Update BR.md

"twenty four" -> "twenty-four"

ryancdickson committed:
Update BR.md

Commit:
Add provision to handle nonces per RFC8954

ryancdickson committed:
Merge pull request #4 from CBonnell/make-ocsp-optional-updates

Add provision to handle nonces per RFC8954

ryancdickson committed:
Update BR.md

Improve readability.

ryancdickson committed:
Update BR.md

ryancdickson committed:
Update BR.md

ryancdickson commented:
Merging and closing PR to help kickoff a second round of public discussion.
